alphagov/paas-compose-broker
# [#149558193] Add bind and unbind ElasticSearch support to Compose broker

## What

This PR introduces the capability to bind and unbind ElasticSearch instances on Compose. This builds upon the previous work to provision and deprovision instances in Compose. At this point we now ave a broker that works with ElasticSearch and is in a suitable state to be released as Private Beta for user feedback. 

As part of this work additional tests have been added for ElasticSearch. These focus on binding to the service, connecting to the service and reading and writing data. 

As part of this work it was necessary to create our own ElasticSearch http client as the 3 clients tested all had issues in our use case, mainly this was due to needing to pass in a certificate in order to connect to Compose.

As part of this PR the MongoDB and ElasticSearch code paths have been broken out into seperate engine functions.

## How to test

If you run the tests locally, please ensure you have ginkgo  >= 1.4.0 installed. As this adds the `--timeout` flag enabling the tests to complete, due to the time taken to crete the instances to test against.

1. Checkout this branch
1. Copy the example catalogue `cp examples/catalog.json .`
1. On your dev environment run `cf push --no-start`
1. Set the username `cf set-env compose-broker USERNAME compose-broker`
1. set the password `cf set-env compose-broker PASSWORD <PASSWORD>`
1. Set the compose API key `cf set-env compose-broker COMPOSE_API_KEY <API Key>` (this is in the credstore)
1. Start the broker `cf start compose-broker`
1. Update the broker catalogue `cf update-service-broker compose-broker compose-broker <PASSWORD> "https://compose-broker.<ENV-NAME>.dev.cloudpipelineapps.digital"`
1. Enable ElasticSearch `cf enable-service-access elastic_search` 
1. Create a service `cf create-service elastic_search tiny <NAME-OF-SERVICE>`
1. Bind the service `cf bind-service <APP-NAME> <NAME-OF-SERVICE>`

## How to merge

Merge this PR before merging https://github.com/alphagov/paas-cf/pull/1000

## Who can merge

Not me, @paroxp,  @henrytk or @combor
